Merge journals & flush journals before lock release

elega opened the pull request:
### What changes are proposed in this pull request?

1. Created a new FileSystemMergeJournalContext to merge journals on a single inode object together before they get flushed.
2. Force flush journals before the locked inode path is closed.

### Why are the changes needed?

Today our current journal are applied different between primary and standby and this creates inconsistencies. The goal of this PR is to mitigate it.
Change 1 reduces the change on a standby master to see partial changes & intermediate metadata. 
Change 2 slightly reduces the race condition and ensures the ordering of journals after FileSystemMergeJournalContext is introduced. 

There will be a following PR to add locks on the standby master side.

### Does this PR introduce any user facing changes?

Please list the user-facing changes introduced by your change, including
  1. change in user-facing APIs
  3. addition or removal of property keys
  4. webui

tcrain commented on core/server/common/src/main/java/alluxio/master/journal/FileSystemMergeJournalContext.java:
```diff
@@ -0,0 +1,99 @@
+    try {
+      mergeAndAppendJournals();
+    } finally {
+      mJournalContext.close();
```
I feel like the parent journal context should be responsible for closing itself?

elega replied:
it closes itself as long as you call mJournalContext.close();. Check this as an example. https://github.com/Alluxio/alluxio/blob/master/core/server/common/src/main/java/alluxio/master/journal/StateChangeJournalContext.java#L47-L55

tcrain commented on core/server/common/src/main/java/alluxio/master/journal/FileSystemMergeJournalContext.java:
```diff
@@ -0,0 +1,99 @@
+                             UnaryOperator<List<JournalEntry>> mergeOperator) {
+    Preconditions.checkNotNull(journalContext, "journalContext");
+    mJournalContext = journalContext;
+    mMergeOperator = mergeOperator;
```
needs a check for null?

elega replied:
sure will add

tcrain commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -573,6 +578,94 @@ public Journaled getDelegate() {
+  public JournalContext createJournalContext() throws UnavailableException {
+    JournalContext context = super.createJournalContext();
+    if (!Configuration.getBoolean(
+        PropertyKey.MASTER_FILE_SYSTEM_MERGE_INODE_JOURNALS
```
Probably all related properties should be static finals? I think it is better that they can't be changed.

elega replied:
k will change

tcrain commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -573,6 +578,94 @@ public Journaled getDelegate() {
+   * @param entries list of journal entries
+   * @return a list of compacted journal entries
+   */
+  public static List<Journal.JournalEntry> mergeFileSystemJournals(
```
I wonder if it is cleaner to merge the journal entires as they are being appended instead of at the end at flush? Also maybe this function can be in a different file.

elega replied:
yeah good idea. will change

tcrain commented on core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java:
```diff
@@ -987,13 +991,22 @@ static void loadFileMetadataInternal(RpcContext rpcContext, LockedInodePath inod
+      RpcContext wrapRpcContext = useInodeSyncMergeJournalContext
+          ? new RpcContext(
+              rpcContext.getBlockDeletionContext(), merger, rpcContext.getOperationContext())
+          : rpcContext;
```
do you think you can cleanup these two statements?

elega replied:
i think it's okay to replace the existing MergeJournalContext with the new one. This nasty code is used to make sure the existing behavior doesn't change when the feature flag is off. wdty

tcrain commented on core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java:
```diff
@@ -273,6 +319,10 @@ public void addNextInode(Inode inode) {
     Preconditions.checkState(!fullPathExists());
     Preconditions.checkState(inode.getName().equals(mPathComponents[mLockList.numInodes()]));
 
+    // We need to flush the pending journals into the writer
```
do you need to flush in removeLastInode() also?

tcrain replied:
and also in traverse()

elega replied:
removeLastInode() needs a flush. will update
traverse() doesn't need that because it actually acquires locks rather than releases/downgrade locks.

tcrain commented:
Do you think you could create some tests specific to this new way of flushing journals. I think maybe the tests for how the flushing works when using InodeTree.getDescendants should be important.

Also it looks like getDescendants takes all locks at the beginning, then releases them all at the end. This could cause an issue with something like the recursive delete, as the in memory journal context can be huge? Can you check this, or see if it is safe to flush midway through?

jiacheliu3 commented on core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java:
```diff
@@ -425,13 +431,18 @@ public LockedInodePath lockInodePath(AlluxioURI uri, LockPattern lockPattern)
    */
-  public LockedInodePath lockInodePath(AlluxioURI uri, LockPattern lockPattern, boolean tryLock)
+  public LockedInodePath lockInodePath(
+      AlluxioURI uri, LockPattern lockPattern, boolean tryLock, JournalContext journalContext
+  )
```
nit: we don't often see this indentation pattern in the codebase

elega replied:
what should i change to though...?

jiacheliu3 commented on core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java:
```diff
@@ -81,6 +85,42 @@ public class LockedInodePath implements Closeable {
   private final ResourceLeakTracker<LockedInodePath> mTracker;
 
+  /**
+   * Keeps a reference of JournalContext and flushes it before the lock is release.
```
```suggestion
   * Keeps a reference of JournalContext and flushes it before the lock is released.
```

jiacheliu3 commented on core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java:
```diff
@@ -81,6 +85,42 @@ public class LockedInodePath implements Closeable {
+   * Keeps a reference of JournalContext and flushes it before the lock is release.
+   * This is used to prevent inconsistency between primary and standby.
+   * A typical write operation in FileSystem looks like:
+   * try (createJournalContext()) {
```
From the other parts of the codebase i see people often put HTML tags in the java doc as below
```suggestion
   * <blockquote><pre>
   * try (createJournalContext()) {
   * ...
   * </pre></blockquote>
```

jiacheliu3 commented on core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java:
```diff
@@ -482,4 +534,16 @@ public void close() {
   }
+
+  private void maybeFlushJournals() {
+    if (Configuration.getBoolean(
```
nit: static final

jiacheliu3 commented on core/common/src/main/java/alluxio/conf/PropertyKey.java:
```diff
@@ -3459,6 +3459,12 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
           .setScope(Scope.MASTER)
           .build();
+  public static final PropertyKey MASTER_FILE_SYSTEM_MERGE_INODE_JOURNALS =
```
Having a switch for an experimental feature is reasonable. However, if we assume the implementation here is all correct, will there still be use cases where we want a `false` here? I guess the answer is no? That leads me to think when we can get rid of this property.

elega replied:
yeah let's release this to the community and find someone dog fooding this first.

tcrain replied:
Maybe should at least set the default to true?

jiacheliu3 reviewed (changes requested):
Added a batch of comments as the initial pass, thanks for the work!

elega commented on core/server/common/src/main/java/alluxio/master/journal/FileSystemMergeJournalContext.java:
```diff
@@ -0,0 +1,99 @@
+      LOG.debug("MergeJournalContext has " + mJournalEntries.size()
+          + " entries, over the limit of " + MAX_ENTRIES);
+    }
+    List<JournalEntry> mergedEntries = mMergeOperator.apply(mJournalEntries);
+    mergedEntries.forEach(mJournalContext::append);
+    mJournalEntries.clear();
```
journal context itself is not thread safe and should not be shared across just threads.

elega commented on core/server/common/src/main/java/alluxio/master/journal/FileSystemMergeJournalContext.java:
```diff
@@ -0,0 +1,99 @@
+@NotThreadSafe
+public final class FileSystemMergeJournalContext implements JournalContext {
+  // It will log a warning if the number of buffered journal entries exceed 100
+  public static final int MAX_ENTRIES = 100;
```
good points. will change.

elega commented on core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java:
```diff
@@ -987,13 +991,22 @@ static void loadFileMetadataInternal(RpcContext rpcContext, LockedInodePath inod
     }
 
+    boolean useInodeSyncMergeJournalContext
+        = !Configuration.getBoolean(
```
done

elega commented:
@tcrain A summary what i've done in the second version to address your comments 
1. Added a new FileSystemJournalEntryMerger to merge the journals instead of implementing a method in FSM. Also journals are merged on the fly now instead of doing it before they are flushed. 
2. Flush the journal in removeLastInode(). No need to flush them in traverse(). See my comments above.
3. Regarding your concern of **getDescendants** creating too many journals in memory in the journal context taking huge memory, I think you concern is 100% valid. What I do here is to add a configurable max journal entries in the MergeJournalContext and we will force flush the journals and clear the journal entry list. Usually only **getDescendants** will create such amount of journals and I think break them up in the middle is reasonable. Though standby master might still see partial state but the possibility is way less than today. 
4. Regarding the test part, I tested the MergeJournalContext with the existing FileSystem fault tolerant testing. In addition to that, I added a FileSystemMasterMergeJournalContextIntegrationTest to test the **getDescendants** case dedicatedly. Also unit tests are added to the JournalMerger and JournalContext as well.
5. Regarding the performance part, I asked @shuaiwuyue to do a performance test with following StressBench test cases and I don't observe any performance regression. 
![image](https://user-images.githubusercontent.com/6771554/185123670-607b4b87-3ded-4494-b359-a27758ceaacc.png)
6. For other nit comments, feel free to comment and @jiacheliu3 and I can work together on them as we are at the same time zone and easier to communicate. 
Thanks a lot!

elega commented on tests/src/test/java/alluxio/server/ft/FileSystemMasterMergeJournalContextIntegrationTest.java:
```diff
@@ -0,0 +1,144 @@
+/*
```
FileSystemMasterMergeJournalContextIntegrationTest @tcrain this test to test the InodeTree.getDescendants case

tcrain replied:
Can you make a simple unit test that makes sure the flush is being called correctly when this is set PropertyKey.MASTER_RECURSIVE_OPERATION_JOURNAL_FORCE_FLUSH_MAX_ENTRIES. For example create a mock that counts the number of flushes and size of journal when it is called during a recursive delete operation please?

elega replied:
sure sure

elega commented on core/server/common/src/main/java/alluxio/master/journal/FileSystemMergeJournalContext.java:
```diff
@@ -0,0 +1,110 @@
+  public void append(JournalEntry entry) {
+    mJournalEntryMerger.add(entry);
+    List<JournalEntry> journalEntries = mJournalEntryMerger.getMergedJournalEntries();
+    if (journalEntries.size() >= MAX_ENTRIES) {
+      LOG.warn("MergeJournalContext has " + journalEntries.size()
+          + " entries, over the limit of " + MAX_ENTRIES + ", triggering force flushing");
+      appendMergedJournals();
```
force flush journals here to address your comment @tcrain

jiacheliu3 replied:
The atomicity of operation may be broken here (as broken as the current alluxio). Let's leave a github issue link here.

tcrain replied:
Yes, I would rather not do it here, and give the responsibility to the operation owner to manually flush the journals as doing it here could make an operation partially visible. Since only the owning operation knows when it is safe to flush the journal it should be the one who has to do this.

Note that the worry is not about making a big recursive operation partially visible and instead about making individual operations partially visible, for example a single create operation. So flushing at a safe point in a recursive operation should be fine. Most file systems, eg POSIX, do not ensure recursive operations are atomic anyway.

For example within the recursive delete operation, it may keep track of how many items it has deleted so far, then flush the journal manually once a certain threshold has been reached. This way you avoid possibly showing a partially item.

elega replied:
okay I made the flush process happen in FSM. @tcrain can you take a look? Thanks!

jiacheliu3 commented on core/server/master/src/main/java/alluxio/master/file/FileSystemJournalEntryMerger.java:
```diff
@@ -0,0 +1,111 @@
+      mJournalEntries.add(entry);
+      mEntriesMap.put(getInodeId(entry), mJournalEntries.size() - 1);
+    }
+    else if (
```
nit: merge into the previous line

jiacheliu3 commented on core/server/master/src/main/java/alluxio/master/file/FileSystemJournalEntryMerger.java:
```diff
@@ -0,0 +1,111 @@
+   * @param entry the new journal entry to add
+   */
+  @Override
+  public void add(Journal.JournalEntry entry) {
```
We are only merging Update into the inodes here right? Any plans for other merges like merging Create and Delete? I guess the intention here is that Create and Update tend to be separate entries due to how those operations are implemented, so that's the only case we care about here?

elega replied:
yeah i don't think we have any file system operation that creates a file and then deletes it. So probably this is okay and we don't have to merge them.

jiacheliu3 reviewed (changes requested):
Mostly LGTM from my perspective. I don't have more comments than this batch. Thanks!

elega commented on core/server/master/src/main/java/alluxio/master/file/FileSystemJournalEntryMerger.java:
```diff
@@ -0,0 +1,111 @@
+          inodeFile.updateFromEntry(entry.getUpdateInodeFile());
+        }
+        mJournalEntries.set(index, inodeFile.toJournalEntry());
+      }
```
we can but this doesn't simplify the code nor did it make the code structure cleaner...

elega commented on core/server/master/src/main/java/alluxio/master/file/FileSystemJournalEntryMerger.java:
```diff
@@ -0,0 +1,111 @@
+    if (entry.hasUpdateInodeDirectory()) {
+      return entry.getUpdateInodeDirectory().getId();
+    }
+    throw new RuntimeException("Unsupported JournalEntry type");
```
done

tcrain reviewed (approved):
Can you also make a comment somewhere that mentions that when this feature is enabled, releasing inode tree locks will flush the entire journal context for the operation, so that whenever a lock is released the operation must be sure that it has made all necessary updates to any journal modifications that should be visible atomically.

jiacheliu3 reviewed (approved):
LGTM, thanks!

elega commented:
alluxio-bot, merge this please

jiacheliu3 commented:
alluxio-bot, merge this please
